File: accounting-frontend/src/app/services/company.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Company, CompanyResponse } from '../models/company.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CompanyService {
  private readonly baseUrl = `${environment.apiUrl}/companies`;

  constructor(private http: HttpClient) {}

  // Get all companies
  getAllCompanies(): Observable<CompanyResponse[]> {
    return this.http.get<CompanyResponse[]>(this.baseUrl);
  }

  // Get all active companies
  getAllActiveCompanies(): Observable<CompanyResponse[]> {
    return this.http.get<CompanyResponse[]>(`${this.baseUrl}/active`);
  }

  // Get company by ID
  getCompanyById(id: string): Observable<CompanyResponse> {
    return this.http.get<CompanyResponse>(`${this.baseUrl}/${id}`);
  }

  // Get company by GSTIN
  getCompanyByGstin(gstin: string): Observable<CompanyResponse> {
    return this.http.get<CompanyResponse>(`${this.baseUrl}/gstin/${gstin}`);
  }

  // Search companies by name
  searchCompaniesByName(name: string): Observable<CompanyResponse[]> {
    return this.http.get<CompanyResponse[]>(`${this.baseUrl}/search`, {
      params: { name }
    });
  }
}

File: src/main/java/com/easy/service/CompanyService.java
package com.easy.service;

import com.easy.entity.CompanyMaster;
import com.easy.repository.CompanyRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class CompanyService {

    private final CompanyRepository companyRepository;

    public List<CompanyMaster> getAllActiveCompanies() {
        log.debug("Fetching all active companies");
        return companyRepository.findByIsActiveTrue();
    }

    public List<CompanyMaster> getAllCompanies() {
        log.debug("Fetching all companies");
        return companyRepository.findAll();
    }

    public Optional<CompanyMaster> getCompanyById(UUID id) {
        log.debug("Fetching company with ID: {}", id);
        return companyRepository.findById(id);
    }

    public CompanyMaster getCompanyByGstin(String gstin) {
        log.debug("Fetching company with GSTIN: {}", gstin);
        return companyRepository.findByGstin(gstin);
    }

    public CompanyMaster getCompanyByPan(String pan) {
        log.debug("Fetching company with PAN: {}", pan);
        return companyRepository.findByPan(pan);
    }

    public List<CompanyMaster> searchCompaniesByName(String companyName) {
        log.debug("Searching companies with name containing: {}", companyName);
        return companyRepository.findByCompanyNameContainingIgnoreCase(companyName);
    }
}
